### DISCUSSION TITLE
Open ssh session in new tab using wezterm ssh?

### DISCUSSION THREAD

**Discussion by adamshand** [wezterm/wezterm#3480] [Q&A] [answered]
Hello,

I've recently discovered WezTerm and really enjoying using it, thanks!

I have a simple script (ssh.sh) which lets me open ssh sessions in new tabs.  With WezTerm I can do this like this:

```
wezterm cli spawn ssh -t example.nz "cd /srv/synapse; exec $SHELL -l"
```

That works great, but I'd prefer to use `wezterm ssh` because that way if I split the tab the new pane opens up on the same host.

I can do that with this command:

```
wezterm ssh -o RequestTTY=yes -o RemoteCommand="cd /srv/synapse; $SHELL -li" example.nz
```

But that opens the new shell in new window rather than a tab.

Is there a way to open ssh sessions in a new tab *and* have splits automatically open on the ssh host?

Thanks for any pointers!

Cheers,
Adam.

**Comment by wez** [comment, marked as answer]
In the latest release, wezterm will automatically populate [ssh_domains](https://wezfurlong.org/wezterm/config/lua/config/ssh_domains.html) from your ssh config.

If you have this in your `~/.ssh/config`

```
Host example.nz
```

you'll have two ssh domains created:

* `SSH:example.nz` - works like `wezterm ssh`
* `SSHMUX:example.nz` - requires wezterm's multiplexer to be installed on the remote

armed with the domain name, you can do:

```console
$ wezterm cli spawn --domain-name SSH:example.nz
```

to use the cli to spawn a new tab in that domain.  [Docs for wezterm cli spawn](https://wezfurlong.org/wezterm/cli/cli/spawn.html).

You can use CTRL-SHIFT-P (by default) to open the [Command Palette](https://wezfurlong.org/wezterm/config/lua/keyassignment/ActivateCommandPalette.html) and fuzzy match `example.nz` and spawn a connection that way, or just fuzzy match on `ssh` to find all your ssh hosts in there.

You can use the CTRL-SHIFT-L (by default) to open the [debug overlay](https://wezfurlong.org/wezterm/config/lua/keyassignment/ShowDebugOverlay.html) and then type in ` wezterm.default_ssh_domains()` and hit enter to see the list it extracted from your ssh configuration.

**Reply by adamshand** [reply]
Woohoo, that works great!  Thanks :-)

Is there any way to make it work for hosts that aren't in `~/.ssh/config`?

**Reply by wez** [reply]
You need to define an ssh domain for the host(s) you want to use. 

```lua
-- populate from ~/.ssh/config
config.ssh_domains = wezterm.default_ssh_domains()

-- add other(s) that are not in the ssh config
table.insert(config.ssh_domains, {
   name = "SSH:some.host",
   remote_address = "some.host"
})
```

See: https://wezfurlong.org/wezterm/multiplexing.html#ssh-domains

**Reply by adamshand** [reply]
I was hoping there was a way for it to work for arbitrary hosts, but no problem.  Thank you!

**Reply by adamshand** [reply]
I'm seeing a problem where the first time I ssh to a host it works as expected, but the second time I get an error.  Not sure if this is a bug or if I'm doing something incorrectly?

```
# the below works as expected
❯ wezterm cli spawn --domain-name SSH:mahoe
1

❯ wezterm cli spawn --domain-name SSH:mahoe
16:23:38.382  ERROR  wezterm > unexpected response Ok(ErrorResponse(ErrorResponse { reason: "Error: Spawning in domain `SSH:mahoe`: TerminalSize { rows: 37, cols: 130, pixel_width: 2860, pixel_height: 1628, dpi: 144 } command=None cwd=None" })); terminating

# the below works as expected
❯ wezterm cli spawn --domain-name SSH:maxx
4

❯ wezterm cli spawn --domain-name SSH:maxx
16:23:50.721  ERROR  wezterm > unexpected response Ok(ErrorResponse(ErrorResponse { reason: "Error: Spawning in domain `SSH:maxx`: TerminalSize { rows: 37, cols: 130, pixel_width: 2860, pixel_height: 1628, dpi: 144 } command=None cwd=None" })); terminating
```

**Reply by wez** [reply]
Should be resolved by fa0f538d361af80e2a459eb0b06e0ba271466be8 and 947d242fdaa22125b9fa01987e88c77d7a842a8d
Those will show up in a nightly build within about an hour.

**Reply by adamshand** [reply]
amazing, thanks!!!!

**Reply by neur1n** [reply]
Where to put `config.ssh_domain = ...` exactly? I tried to put them right in _.wezterm.lua_ but it complains about undefined variable. Thanks in advaned.

**Reply by arthurgeek** [reply]
When using `wezterm ssh host`, it uses the ssh agent and don't ask for my ssh passphrase, but `wezterm cli spawn --domain-name SSH:host` do ask for the passphrase.

**Reply by codemedic** [reply]
When there are 100s or 1000s of hosts to log on to, it doesn't make sense to add them all to `.ssh/config`. I am in that situation; the way I have my SSH configured is to use wildcard based rules that setup multiple jump rules etc. OpenSSH CLI works really well in this fashion.

Any chance to spawn the OpenSSH command rather than using the rust implementation?

What I was hoping to achieve was to use a key mapping to invoke a lua function that somehow asks for hostname to connect to and opens a new tab or pane; a workaround for [#5800](https://github.com/wez/wezterm/issues/5800)
